akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala

```diff
@@ -523,8 +514,8 @@ object Flow {
    *
    * '''Cancels when''' downstream cancels
    */
-  def lazyInit[I, O, M](flowFactory: I ⇒ Future[Flow[I, O, M]], fallback: () ⇒ M): Flow[I, O, M] =
-    Flow.fromGraph(new LazyFlow[I, O, M](flowFactory, fallback))
+  def lazyInit[I, O, M](flowFactory: I ⇒ Future[Flow[I, O, M]]): Flow[I, O, Future[Option[M]]] =
```

Comment:
If we change the signature like this then please do keep the old method and mark it as deprecated, this way we can phase it out without breaking binary compatibility for users

Reply:
Ok. I  can add a method implementing the old signature. The implementation does always use the value created by the `fallback` function as the materialized value. (There is no other way to implement that signature.)

Reply:
Yeah, it will have to be such crippled impl

Reply:
There is a slight complication: method overloading does not work well with function parameter derivation, i.e. the compiler does not infer that the `flowFactory` takes an `I`. The parameter type has to specified manually: `Flow.lazyInit[Int, Int, Mat]((_: Int) =>...)` 

I wonder if it is really that useful that the `flowFactory` is a function that gets the first element as input. In addition, it would be more convenient (type derivation wise) if a call by name argument or a function without parameters is used:

```
def lazyInit[In, Out, Mat](flowFactory: => Future[Flow[In, Out, Mat]]]): ...
```

This would allow to call `Flow.lazyInit(someFlow)` without having to specify any types.

Reply:
One more note: the mentioned overloading issue would hit all usage sites of the deprecated method if the new method has the same name. Maybe the new method should get a new name, too?

Reply:
Yet another note: `Source` distinguishes `Source.lazily` and `Source.lazilyAsync` the latter returning a `Future` from the factory. Therefore a suitable name for the new method might be `Flow.lazyInitAsync`. After its deprecation phase the original method may be replaced by a new implementation `def lazyInit[In, Out, Mat](flowFactory: () => Flow[In, Out, Mat]`): ...`